### DISCUSSION TITLE
getimage not capturing background and certain window contents

### DISCUSSION THREAD

**Discussion by ghost** [jezek/xgb#20] [Q&A] [answered]
I am using xproto.Getimage to try and capture the state of the screen. But when i do certain windows have no background and the desktop wallpaper is missing/completely transparent. I was unable to find a solution online so I wanted to  ask here.

the line of code in question:
```go
imgCookie := xproto.GetImage(Xconn, xproto.ImageFormatZPixmap, xproto.Drawable(screen.Root), x, y, width, height, 0xffffffff)
```

**Comment by jezek** [comment]
Are you on wayland or X11? Maybe wayland has an issue to give an potentially unsecure X11 app access to background and other wayland apps.

Note: I found another way to [capture screen using SHM](https://github.com/BurntSushi/xgb/issues/28) in BurntSushi/xgb issues. It's low probability that it works better than standard xproto.GetImage, but maybe it's worth a try.

**Comment by ghost** [comment]
Its not a wayland/x11 issue I think that it is an encoding issue, when I convert the rgba array to image.Image then encode as a png certain aspects are transparent that are not supposed to be. But when i do the same thing and encode as a jpeg the image comes out as its supposed to. You could say this is problem solved but i want make the image comes out as its supposed to at the rgba level since that is what i need to handle in my actual program.

so now the question is what is jpeg.Encode doing to the image.Image and how do we do this to the rgba array?

Below is a test program and resulting images.

```go
package main

import (
	"fmt"
	"image"
	"image/jpeg"
	"image/png"
	"os"

	"github.com/jezek/xgb"
	"github.com/jezek/xgb/xproto"
)

func main() {
	X, err := xgb.NewConn()
	if err != nil {
		fmt.Println(err)
		return
	}

	reply, err := getRect(X, 0, 0, 1920, 1080)
	if err != nil {
		fmt.Println(err)
		return
	}

	rgba_IMG := image.NewRGBA(image.Rect(0, 0, 1920, 1080))
	rgba_IMG.Pix = make([]uint8, reply.Length*4)

	copy(rgba_IMG.Pix, reply.Data)
	bgra2rgba(rgba_IMG.Pix)

	rgba_IMG.SubImage(image.Rect(0, 0, 1920, 1080))
	saveImage(rgba_IMG, "screenshot.png", "png")
	saveImage(rgba_IMG, "screenshot.jpeg", "jpeg")
}
func getRect(Xconn *xgb.Conn, x int16, y int16, width uint16, height uint16) (*xproto.GetImageReply, error) {
	screen := xproto.Setup(Xconn).DefaultScreen(Xconn)

	imgCookie := xproto.GetImage(Xconn, xproto.ImageFormatZPixmap, xproto.Drawable(screen.Root), x, y, width, height, 0xffffffff)
	imgReply, err := imgCookie.Reply()
	if err != nil {
		return nil, err
	}

	return imgReply, nil
}

func saveImage(img image.Image, filename, filetype string) error {

	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	if filetype == "png" {
		err = png.Encode(file, img)
	} else if filetype == "jpeg" {
		err = jpeg.Encode(file, img, nil)
	}
	if err != nil {
		return err
	}

	return nil
}

func bgra2rgba(data []byte) {
	for i := 0; i < len(data); i += 4 {
		data[i], data[i+2] = data[i+2], data[i]
	}
}

``` 
png:
(notice file explore has a transparent background also)
![screenshot](https://github.com/jezek/xgb/assets/148909146/58d78e5f-796e-462e-b4d6-1e6789ae54e9)

jpeg:
![screenshot](https://github.com/jezek/xgb/assets/148909146/35ccee55-aeaf-4079-becd-9e70637fac4d)

**Comment by ghost** [comment, marked as answer]
I have found the solution in bgra2rgba function i now set the alpha layer to max (255), this might be a bug or might just be x11 i don't know, but it took me a while since i was setting transparency to max in the actual GetImage function.

updated function:
```go
func bgra2rgba(data []byte) {
	for i := 0; i < len(data); i += 4 {
		data[i], data[i+2] = data[i+2], data[i]
		data[i+3] = 255
	}
}
```

Hope this helps someone!
